src/rhsmlib/cloud/providers/aws.py

```diff
@@ -0,0 +1,122 @@
+#
+# Red Hat trademarks are not licensed under GPLv2. No permission is
+# granted to use or replicate Red Hat trademarks that are incorporated
+# in this software or its documentation.
+#
+
+"""
+This is module implementing detector and metadata collector of virtual machine running on AWS
+"""
+
+from rhsmlib.cloud.detector import CloudDetector
+
+
+class AWSCloudDetector(CloudDetector):
+    """
+    Detector of cloud machine
+    """
+
+    ID = 'aws'
+
+    def __init__(self, hw_info):
+        """
+        Initialize instance of AWSCloudDetector
+        """
+        super(AWSCloudDetector, self).__init__(hw_info)
+
+    def is_vm(self):
+        """
+        Is system running on virtual machine or not
+        :return: True, when machine is running on VM; otherwise return False
+        """
+        return super(AWSCloudDetector, self).is_vm()
+
+    def is_running_on_cloud(self):
+        """
+        Try to guess if cloud provider is AWS using collected hardware information (output of dmidecode,
+        virt-what, etc.)
+        :return: True, when we detected sign of AWS in hardware information; Otherwise return False
+        """
+
+        # The system has to be VM
+        if self.is_vm() is False:
+            return False
+        # This is valid for AWS systems using Xen
+        if 'dmi.bios.version' in self.hw_info and 'amazon' in self.hw_info['dmi.bios.version']:
+            return True
+        # This is valid for AWS systems using KVM
+        if 'dmi.bios.vendor' in self.hw_info and 'Amazon EC2' in self.hw_info['dmi.bios.vendor']:
+            return True
+        # Try to get output from virt-what
+        if 'virt.host_type' in self.hw_info and 'aws' in self.hw_info['virt.host_type']:
+            return True
+        # In other cases return False
+        return False
+
+    def is_likely_running_on_cloud(self):
+        """
+        Return non-zero value, when the machine is virtual machine and it is running on kvm/xen and
+        some Amazon string can be found in output of dmidecode
+        :return: Float value representing probability that vm is running on AWS
+        """
+        probability = 0.0
+
+        # When the machine is not virtual machine, then there is probably zero chance that the machine
+        # is running on AWS
+        if self.is_vm() is False:
+            return 0.0
+
+        # We know that AWS uses mostly KVM and it uses Xen in some cases
+        if 'virt.host_type' in self.hw_info:
+            # It seems that KVM is used more ofter
+            if 'kvm' in self.hw_info['virt.host_type']:
+                probability += 0.3
+            elif 'xen' in self.hw_info['virt.host_type']:
+                probability += 0.2
+
+        # Try to find "Amazon EC2", "Amazon" or "AWS" keywords in output of dmidecode
+        found_amazon = False
+        found_amazon_ec2 = False
+        found_aws = False
+        for hw_item in self.hw_info.values():
```

Comment:
Should we expand the list of items we are checking for to include checking if the /sys/hypervisor/uuid file has "ec2" at the beginning?

https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/identify_ec2_instances.html

Reply:
We could do it, but wow. This is so weak sign that machine is running on AWS, because the probability that some other machine has ec2 at the beginning of system uuid is quite high. It is interesting that the aws_instance_id is included in the file: /sys/devices/virtual/dmi/id/board_asset_tag

Reply:
If it is a very likely false positive then I agree we should leave it out.